File: tools/encode_kitchen_walkthrough.m
#import <Foundation/Foundation.h>
#import <AVFoundation/AVFoundation.h>
#import <ImageIO/ImageIO.h>
#import <CoreGraphics/CoreGraphics.h>
#import <CoreVideo/CoreVideo.h>

// Build: clang -fobjc-arc tools/encode_kitchen_walkthrough.m \
//   -framework Foundation -framework AVFoundation -framework ImageIO \
//   -framework CoreGraphics -framework CoreVideo -framework CoreMedia \
//   -o /private/tmp/encode_kitchen_walkthrough

static void fail(NSString *message) {
    fprintf(stderr, "%s\n", message.UTF8String);
    exit(1);
}

int main(int argc, const char *argv[]) {
    @autoreleasepool {
        if (argc != 4) fail(@"Usage: encode_kitchen_walkthrough <frames-directory> <output.mp4> <fps>");
        NSURL *folder = [NSURL fileURLWithPath:[NSString stringWithUTF8String:argv[1]] isDirectory:YES];
        NSURL *output = [NSURL fileURLWithPath:[NSString stringWithUTF8String:argv[2]]];
        int fps = atoi(argv[3]);
        if (fps < 1) fail(@"FPS must be positive");
        NSError *error = nil;
        NSArray<NSURL *> *contents = [[NSFileManager defaultManager] contentsOfDirectoryAtURL:folder
            includingPropertiesForKeys:nil options:0 error:&error];
        if (!contents) fail(error.localizedDescription);
        NSPredicate *framesOnly = [NSPredicate predicateWithBlock:^BOOL(NSURL *url, NSDictionary *bindings) {
            return [url.lastPathComponent hasPrefix:@"frame_"] &&
                [url.pathExtension.lowercaseString isEqualToString:@"jpg"];
        }];
        NSArray<NSURL *> *frames = [[contents filteredArrayUsingPredicate:framesOnly]
            sortedArrayUsingComparator:^NSComparisonResult(NSURL *a, NSURL *b) {
                return [a.lastPathComponent compare:b.lastPathComponent];
            }];
        if (frames.count == 0) fail(@"No JPEG frames found");

        CGImageSourceRef firstSource = CGImageSourceCreateWithURL((__bridge CFURLRef)frames[0], NULL);
        CGImageRef first = firstSource ? CGImageSourceCreateImageAtIndex(firstSource, 0, NULL) : NULL;
        if (!first) fail(@"The first frame cannot be read");
        size_t width = CGImageGetWidth(first), height = CGImageGetHeight(first);
        CGImageRelease(first);
        CFRelease(firstSource);

        [[NSFileManager defaultManager] removeItemAtURL:output error:nil];
        AVAssetWriter *writer = [[AVAssetWriter alloc] initWithURL:output fileType:AVFileTypeMPEG4 error:&error];
        if (!writer) fail(error.localizedDescription);
        NSDictionary *settings = @{
            AVVideoCodecKey: AVVideoCodecTypeH264,
            AVVideoWidthKey: @(width),
            AVVideoHeightKey: @(height),
            AVVideoCompressionPropertiesKey: @{
                AVVideoAverageBitRateKey: @5000000,
                AVVideoProfileLevelKey: AVVideoProfileLevelH264HighAutoLevel
            }
        };
        AVAssetWriterInput *input = [AVAssetWriterInput assetWriterInputWithMediaType:AVMediaTypeVideo
            outputSettings:settings];
        input.expectsMediaDataInRealTime = NO;
        if (![writer canAddInput:input]) fail(@"Cannot add the video track");
        [writer addInput:input];
        NSDictionary *attributes = @{
            (id)kCVPixelBufferPixelFormatTypeKey: @(kCVPixelFormatType_32BGRA),
            (id)kCVPixelBufferWidthKey: @(width),
            (id)kCVPixelBufferHeightKey: @(height),
            (id)kCVPixelBufferCGImageCompatibilityKey: @YES,
            (id)kCVPixelBufferCGBitmapContextCompatibilityKey: @YES
        };
        AVAssetWriterInputPixelBufferAdaptor *adaptor =
            [AVAssetWriterInputPixelBufferAdaptor assetWriterInputPixelBufferAdaptorWithAssetWriterInput:input
                sourcePixelBufferAttributes:attributes];
        if (![writer startWriting]) fail(writer.error.localizedDescription);
        [writer startSessionAtSourceTime:kCMTimeZero];
        CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();

        for (NSUInteger index = 0; index < frames.count; index++) {
            @autoreleasepool {
                while (!input.readyForMoreMediaData) [NSThread sleepForTimeInterval:0.005];
                CGImageSourceRef source = CGImageSourceCreateWithURL((__bridge CFURLRef)frames[index], NULL);
                CGImageRef image = source ? CGImageSourceCreateImageAtIndex(source, 0, NULL) : NULL;
                if (!image || CGImageGetWidth(image) != width || CGImageGetHeight(image) != height)
                    fail([NSString stringWithFormat:@"Bad frame: %@", frames[index].path]);
                CVPixelBufferRef buffer = NULL;
                if (CVPixelBufferPoolCreatePixelBuffer(kCFAllocatorDefault, adaptor.pixelBufferPool, &buffer) != kCVReturnSuccess)
                    fail(@"Cannot allocate a video frame");
                CVPixelBufferLockBaseAddress(buffer, 0);
                CGContextRef context = CGBitmapContextCreate(CVPixelBufferGetBaseAddress(buffer), width, height,
                    8, CVPixelBufferGetBytesPerRow(buffer), colorSpace,
                    kCGImageAlphaPremultipliedFirst | kCGBitmapByteOrder32Little);
                if (!context) fail(@"Cannot draw a video frame");
                CGContextDrawImage(context, CGRectMake(0, 0, width, height), image);
                CGContextRelease(context);
                CVPixelBufferUnlockBaseAddress(buffer, 0);
                CMTime timestamp = CMTimeMake((int64_t)index, fps);
                if (![adaptor appendPixelBuffer:buffer withPresentationTime:timestamp])
                    fail([NSString stringWithFormat:@"Cannot append frame %lu: %@", (unsigned long)index,
                        writer.error.localizedDescription ?: @"unknown error"]);
                CVPixelBufferRelease(buffer);
                CGImageRelease(image);
                CFRelease(source);
                if (index % 240 == 0) printf("Encoded %lu/%lu frames\n", (unsigned long)index,
                    (unsigned long)frames.count);
            }
        }
        CGColorSpaceRelease(colorSpace);
        [input markAsFinished];
        dispatch_semaphore_t done = dispatch_semaphore_create(0);
        [writer finishWritingWithCompletionHandler:^{ dispatch_semaphore_signal(done); }];
        dispatch_semaphore_wait(done, DISPATCH_TIME_FOREVER);
        if (writer.status != AVAssetWriterStatusCompleted) fail(writer.error.localizedDescription);
        printf("Wrote %s: %lu frames, %zux%zu, %d fps\n", output.path.UTF8String,
            (unsigned long)frames.count, width, height, fps);
    }
    return 0;
}

File: tools/encode_kitchen_walkthrough.swift
import Foundation
import AVFoundation
import ImageIO
import CoreGraphics
import CoreVideo

guard CommandLine.arguments.count == 4 else {
    fputs("Usage: swift encode_kitchen_walkthrough.swift <frames-directory> <output.mp4> <fps>\n", stderr)
    exit(2)
}

let source = URL(fileURLWithPath: CommandLine.arguments[1], isDirectory: true)
let output = URL(fileURLWithPath: CommandLine.arguments[2])
let fps = Int(CommandLine.arguments[3]) ?? 24
let files = try FileManager.default.contentsOfDirectory(at: source, includingPropertiesForKeys: nil)
    .filter { $0.lastPathComponent.hasPrefix("frame_") && $0.pathExtension == "jpg" }
    .sorted { $0.lastPathComponent < $1.lastPathComponent }
guard !files.isEmpty, let firstSource = CGImageSourceCreateWithURL(files[0] as CFURL, nil),
      let first = CGImageSourceCreateImageAtIndex(firstSource, 0, nil) else {
    fputs("No readable JPEG frames\n", stderr)
    exit(3)
}

let width = first.width
let height = first.height
try? FileManager.default.removeItem(at: output)
let writer = try AVAssetWriter(outputURL: output, fileType: .mp4)
let settings: [String: Any] = [
    AVVideoCodecKey: AVVideoCodecType.h264,
    AVVideoWidthKey: width,
    AVVideoHeightKey: height,
    AVVideoCompressionPropertiesKey: [
        AVVideoAverageBitRateKey: 5_000_000,
        AVVideoProfileLevelKey: AVVideoProfileLevelH264HighAutoLevel
    ]
]
let input = AVAssetWriterInput(mediaType: .video, outputSettings: settings)
input.expectsMediaDataInRealTime = false
guard writer.canAdd(input) else { fatalError("Cannot add video input") }
writer.add(input)
let attributes: [String: Any] = [
    kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA,
    kCVPixelBufferWidthKey as String: width,
    kCVPixelBufferHeightKey as String: height,
    kCVPixelBufferCGImageCompatibilityKey as String: true,
    kCVPixelBufferCGBitmapContextCompatibilityKey as String: true
]
let adaptor = AVAssetWriterInputPixelBufferAdaptor(assetWriterInput: input, sourcePixelBufferAttributes: attributes)
guard writer.startWriting() else { fatalError("Cannot start MP4: \(writer.error?.localizedDescription ?? "unknown")") }
writer.startSession(atSourceTime: .zero)

for (index, file) in files.enumerated() {
    while !input.isReadyForMoreMediaData { Thread.sleep(forTimeInterval: 0.005) }
    try autoreleasepool {
        guard let source = CGImageSourceCreateWithURL(file as CFURL, nil),
              let image = CGImageSourceCreateImageAtIndex(source, 0, nil),
              image.width == width, image.height == height else {
            throw NSError(domain: "KitchenVideo", code: 4, userInfo: [NSLocalizedDescriptionKey: "Unreadable frame: \(file.path)"])
        }
        var buffer: CVPixelBuffer?
        guard let pool = adaptor.pixelBufferPool,
              CVPixelBufferPoolCreatePixelBuffer(kCFAllocatorDefault, pool, &buffer) == kCVReturnSuccess,
              let pixelBuffer = buffer else {
            throw NSError(domain: "KitchenVideo", code: 5, userInfo: [NSLocalizedDescriptionKey: "Could not allocate pixel buffer"])
        }
        CVPixelBufferLockBaseAddress(pixelBuffer, [])
        defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, []) }
        guard let base = CVPixelBufferGetBaseAddress(pixelBuffer),
              let context = CGContext(data: base, width: width, height: height,
                                      bitsPerComponent: 8, bytesPerRow: CVPixelBufferGetBytesPerRow(pixelBuffer),
                                      space: CGColorSpaceCreateDeviceRGB(),
                                      bitmapInfo: CGImageAlphaInfo.premultipliedFirst.rawValue | CGBitmapInfo.byteOrder32Little.rawValue) else {
            throw NSError(domain: "KitchenVideo", code: 6, userInfo: [NSLocalizedDescriptionKey: "Could not draw frame"])
        }
        context.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))
        let time = CMTime(value: Int64(index), timescale: Int32(fps))
        guard adaptor.append(pixelBuffer, withPresentationTime: time) else {
            throw NSError(domain: "KitchenVideo", code: 7, userInfo: [NSLocalizedDescriptionKey: "Could not append frame \(index): \(writer.error?.localizedDescription ?? "unknown")"])
        }
    }
    if index % 240 == 0 { print("Encoded \(index)/\(files.count) frames") }
}
input.markAsFinished()
let finished = DispatchSemaphore(value: 0)
writer.finishWriting { finished.signal() }
finished.wait()
guard writer.status == .completed else { fatalError("MP4 export failed: \(writer.error?.localizedDescription ?? "unknown")") }
print("Wrote \(output.path): \(files.count) frames, \(width)x\(height), \(fps) fps")
